src/main/java/picard/sam/util/SamComparison.java

```diff
@@ -22,40 +26,150 @@
     }
 
-    public void printReport() {
-        System.out.println("Match\t" + mappingsMatch);
-        System.out.println("Differ\t" + mappingsDiffer);
-        System.out.println("Unmapped_both\t" + unmappedBoth);
-        System.out.println("Unmapped_left\t" + unmappedLeft);
-        System.out.println("Unmapped_right\t" + unmappedRight);
-        System.out.println("Missing_left\t" + missingLeft);
-        System.out.println("Missing_right\t" + missingRight);
+    private void countLenientDuplicateMarkingDifferences() {
+        /* Count duplicate marking differences allowing swaps within duplicate sets.  Should only be used when lenientDup == true.
+           Reads which have different duplicate markings will have been saved in markDuplicatesCheckLeft and markDuplicatesCheckRight
+           in catalogDuplicateDifferences.
+         */
+
+        if (!lenientDup) {
+            throw new PicardException("Should only use countLenientDuplicateMarkingDifferences when in lenient duplicate marking mode.");
+        }
+
+        final DuplicateSetIterator duplicateSetLeftIterator = new DuplicateSetIterator(markDuplicatesCheckLeft.iterator(), reader1.getFileHeader(), true);
+        final DuplicateSetIterator duplicateSetRightIterator = new DuplicateSetIterator(markDuplicatesCheckRight.iterator(), reader2.getFileHeader(), true);
+        /* Iterate through duplicateSetIterators.  This will iterate through all reads with differing duplicate markings in left and right files.  Passing through the first
+         * iterator (i.e. left file), for each duplicate set we create a set of duplicate reads that the representative read could "swap" with.  On the pass through the second iterator, for each
+         * duplicate read, we look for the reads it is allowed to swap with in the duplicate set from the first iterator.  If the allowed swap can also be performed in this duplicate set,
+         * we do not count the read or the read it could be swapped with as having mismatched duplicates.  Note however, that once a swap is allowed, the reads swapped cannot be used for any
+         * further swaps.
+         */
+        final HashMap<String, Set<String>> allowedSwapsMap = new HashMap<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetLeftIterator)) {
+            // do not want to redo duplicate marking, want to keep marks assigned in input files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            final Set<String> duplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                    .collect(Collectors.toSet());
+            final Map<String, Set<String>> repSwapsMap = records.stream().filter(r -> !r.getDuplicateReadFlag())
+                    .collect(Collectors.toMap(SAMRecord::getReadName, r -> duplicateRecordNames, (oldV, newV) -> oldV, HashMap::new));
+            allowedSwapsMap.putAll(repSwapsMap);
```

Comment:
BTW, the test file "dup3" claims to be coordinate sorted, but it isn't actually in order. I would suggest sorting and also validating the test files to make sure they're valid, and also naming them with a name that indicates what they contain (that can be hard to do but even something like "twoDuplicateSets.sam", "threeDuplicateSets.sam").

Reply:
I believe the logic does work here, though I agree it can lead to a confusing state.  It is actually fine to have `allowedSwapsMap` have empty values.  `allowedSwapsMap` is storing the possible duplication swaps that can occur in the left file, with the keys being the representative reads and the values being the set of duplicates each representative read could be swapped with.  When an entry in `allowedSwapsMap` has an empty value, it just means that is it a representative read that has no duplicates that we will allow it to swap with.  Similarly, when there is no representative read in a duplicate set, the duplicate reads will not be added as a value to `allowedSwapsMap`, since there is no representative read with which they are allowed to swap.

For the example of dup3 as the left file and dup1 as the right,  the 4 `DuplicateSets` correspond to the 4 reads (2 read pairs) which have different duplicate marking between dup3 and dup1 (`DuplicateSetIterator` gives singleton duplicate sets when a read has no duplicates in the iterator).  Of those, 2 reads (1 read pair) are marked as non-duplicates, and so included in `allowedSwapsMap` with an empty value.  However, this read pair is not allowed to swap and be saved, because of it's empty value in the map, and so both read pairs are counted as being marked differently between the two files, as expected.

While this works, it is not actually necessary to perform these checks on singleton duplicate sets, so I have added a check on the size of the duplicate set in each loop to avoid unnecessary/confusing computation.